### PR TITLE
Maximum cardinality matching in bipartite graphs.

### PR DESCRIPTION
**Pull request by jfinkels**
Adds the Hopcroft--Karp algorithm for finding a maximum cardinality
matching in bipartite graphs.

I wasn't sure whether to put this in `networkx.algorithms.matching` or in `networkx.algorithms.bipartite`; I chose the latter.

### CODE REVIEW

**Comment by jfinkels** [comment]
The previous suggested changes failed on Python2.7. I have updated the pull request to resolve that issue.

**Comment by jfinkels** [comment]
The Travis CI errors seem to be from an unrelated unit test, `test_timing`. If you examine the build log and search for the string `Hopcroft--Karp`, you'll see that the test passes.

**Comment by hagberg** [comment]
Yes, you are right.  Those tests are not related to your code but instead in the timing tests.  @dschult can you look at this?  Do we need to run these tests regularly or are they primarily for evaluation performance  when suggesting changes?

**Comment by hagberg** [comment]
@jfinkels I took at look at the code and I see it follows the wikipedia article as you mention.  I was looking at the Eppstein implementation http://www.ics.uci.edu/~eppstein/PADS/BipartiteMatching.py which is quite different.  Do you have comments on the differences?

**Comment by jfinkels** [comment]
I'm not familiar enough with the algorithm to say, but I trust that Professor Eppstein's implementation is as good if not better than mine (or Wikipedia's). It seems to be the same general idea done in a slightly different way, but it's hard for me to say for certain without spending a lot of time trying to interpret the code in both implementations.

I can add his implementation as well, choose a reasonable default implementation to present to the user, and then allow the user to explicitly choose one or the other in case one is more efficient.

**Comment by jfinkels** [comment]
I've just updated this pull request again; this time I have two implementations of the Hopcroft--Karp maximum matching algorithm for bipartite graphs: one from Wikipedia and one from David Eppstein's PADS. I've created a module-level attribute `networkx.bipartite.maximum_matching` which is simply an alias for one of the algorithms; users who don't care about implementation should use this one. Finally, I have added an additional function, which computes the minimum vertex cover corresponding to the maximum matching computed by (either of) the maximum matching functions.

File: networkx/algorithms/bipartite/matching.py [hunk 1]
```diff
@@ -0,0 +1,422 @@
...
+    # invocation of the search functions.
+    def breadth_first_search():
+        for v in left:
+            if leftmatches[v] is None:
+                distances[v] = 0
+                queue.append(v)
+            else:
+                distances[v] = INFINITY
+        distances[None] = INFINITY
+        while len(queue) > 0:
```
Comment (ysitu) [inline networkx/algorithms/bipartite/matching.py]: `while queue:`

File: networkx/algorithms/bipartite/matching.py [hunk 2]
```diff
@@ -0,0 +1,422 @@
...
+            v = queue.pop(0)
```
Comment (ysitu) [inline networkx/algorithms/bipartite/matching.py]: Use `collections.deque` and `popleft` to avoid quadratic behavior.

**Comment by jfinkels** [comment]
I have made the changes suggested by @ysitu; specifically using an instance of `collections.deque` instead of a plain list to use as a queue.

File: networkx/algorithms/bipartite/matching.py [hunk 3]
```diff
@@ -0,0 +1,423 @@
...
+            # since that information can be inferred from the matching
+            # dictionary.
+            return matching
+
+        # recursively search backward through layers to find alternating paths
+        # recursion returns true if found path, false otherwise
+        def recurse(v):
+            if v in preds:
+                L = preds[v]
+                del preds[v]
```
Comment (ysitu) [inline networkx/algorithms/bipartite/matching.py]: `L = preds.pop(v)`

File: networkx/algorithms/bipartite/matching.py [hunk 4]
```diff
@@ -0,0 +1,423 @@
...
+                for u in L:
+                    if u in pred:
+                        pu = pred[u]
+                        del pred[u]
```
Comment (ysitu) [inline networkx/algorithms/bipartite/matching.py]: `pu = pred.pop(u)`

**Comment by jfinkels** [comment]
@ysitu Fixed again.

**Pull request merged**